pkg/blockbuilder/partitions.go

```diff
@@ -0,0 +1,73 @@
+package blockbuilder
+
+import "slices"
+
+// partitions represent a queue of partitions currently assigned to the consumer.
+type partitions struct {
+	list []int32
+	cur  int
+}
+
+// next returns the next item in the queue of assigned partitions.
+func (p *partitions) next() int32 {
+	if len(p.list) == 0 {
+		return -1
+	}
+
+	if p.cur >= len(p.list) {
+		return -1
+	}
+
+	i := p.cur
+	p.cur++
+
+	return p.list[i]
+}
+
+// update replaces the queue with a new list of assigned partitions.
+// It doesn't reset the queue's cursor. That is, the later call to next will return the last active item
+// if the item exists in the new list.
+func (p *partitions) update(list []int32) {
```

Comment:
So, if I am understanding it right, when update() happens during an on-going consumption, say: if the BB was consuming parts 1 3 5, and it was at 3 now. part 1 and 5 were removed and part 2 and 4 were added to this BB. So the partitions will get updated as 3 2 4 with cursor at 3. So the BB will continue consumption and always consume the newly added partitions (2 and 4 here). Right?

I have one concern here. A race is possible: if there is a BB2 that just got done with consumption cycle and gets assigned part 5 from the above BB, now part 5 will lag for a cycle because the above BB had not processed it yet.

I think we should check if a partition is up to date with the BB's consumption cycle by checking its last commit, and do some consumption if it is lagging, when a partition is newly assigned to a BB (this can be in a follow up PR)

Reply:
I think you're right about the particular case. But I'd see that as an optimization to explore later.

There is a similar scenario possible, when a BB instance is assigned partitions 1, 3, 5. There is a chance that 3 and 5 won't have anything to consume because the ingesters that own these partitions were downscaled.

If BB finishes its cycle before ingesters are scaled up again, the partitions 3 and 5 will end up lagging for one cycle.